api/tests/unit/app_analytics/test_tasks.py

```diff
@@ -0,0 +1,318 @@
+from datetime import datetime, timedelta
+
+import pytest
+from app_analytics.models import (
+    APIUsageBucket,
+    APIUsageRaw,
+    FeatureEvaluationBucket,
+    FeatureEvaluationRaw,
+    Resource,
+)
+from app_analytics.tasks import (
+    populate_api_usage_bucket,
+    populate_feature_evaluation_bucket,
+    track_feature_evaluation,
+    track_request,
+)
+from django.conf import settings
+from django.utils import timezone
+
+if "analytics" not in settings.DATABASES:
+    pytest.skip(
+        "Skip test if analytics database is configured", allow_module_level=True
+    )
+
+
+def _create_api_usage_event(environment_id: str, when: datetime):
+    event = APIUsageRaw.objects.create(
+        environment_id=environment_id,
+        host="host1",
+        resource=Resource.FLAGS,
+    )
+    # update created_at
+    event.created_at = when
+    event.save()
+
+    return event
+
+
+@pytest.mark.freeze_time("2023-01-19T09:09:47.325132+00:00")
+@pytest.mark.django_db(databases=["analytics"])
+def test_populate_api_usage_bucket_15m_bucket(freezer):
+    # Given
+    environment_id = 1
+    bucket_size = 15
+    now = timezone.now()
+    # let's create events at every 1 minutes
+    # for the last two hours, i.e: from 9:09:47 to 7:10:47
+    for i in range(60 * 2):
+        _create_api_usage_event(environment_id, now - timezone.timedelta(minutes=1 * i))
+        # create events in some other environments as well - just to make sure
+        # we don't aggregate them in the same environment
+        _create_api_usage_event(999, now - timezone.timedelta(minutes=1 * i))
+
+    # Next, let's go 1 hr back in the past and run this
+    freezer.move_to(timezone.now() - timezone.timedelta(hours=1))
+    populate_api_usage_bucket(bucket_size, run_every=timedelta(minutes=60))
+
+    # Then - we should have four buckets
```

Comment:
Shouldn't we have 5 buckets? I think we agreed that we would do `(run_every // bucket size) + 1` ?

Reply:
My concern is that we could miss events on the join between processor intervals. So, for example if the task to populate the buckets runs at say 10:15:01, it will start processing (backwards) from the bucket that is bounded by 10 and 10:15. If there is a raw event that happens at 10:14:59 but doesn't get written to the RawEvent table until 10:15:01 then it could be missed. 

As discuss, it sounds like this is not an issue as the event would be written with a date of 10:15:01 and hence will be counted in the next bucket.

The downside to this is that we are not necessarily accurately tracking the request time but I don't think that this matters in our use case. Do you agree @dabeeeenster ?